src/nanoemoji/write_font.py

```diff
@@ -310,23 +331,95 @@ def _glyf_ufo(ufo, color_glyphs):
             assert component.name == color_glyph.glyph_name
 
 
-def _colr_paint(colr_version: int, paint: Paint, palette: Sequence[Color]):
+def _colr_layer(
+    colr_version: int, layer_glyph_name: str, paint: Paint, palette: Sequence[Color]
+):
     # For COLRv0, paint is just the palette index
     # For COLRv1, it's a data structure describing paint
     if colr_version == 0:
         # COLRv0: draw using the first available color on the glyph_layer
         # Results for gradients will be suboptimal :)
         color = next(paint.colors())
-        return palette.index(color)
+        return (layer_glyph_name, palette.index(color))
 
     elif colr_version == 1:
-        # COLRv1: solid or gradient paint
-        return paint.to_ufo_paint(palette)
+        # COLRv1: layer is graph of (solid, gradients, etc.) paints.
+        # Root node is always a PaintGlyph (format 4) for now.
+        # TODO Support more paints (PaintTransform, PaintColorGlyph, PaintComposite)
+        return PaintGlyph(layer_glyph_name, paint).to_ufo_paint(palette)
 
     else:
         raise ValueError(f"Unsupported COLR version: {colr_version}")
 
 
+def _inter_glyph_reuse_key(painted_layer: PaintedLayer) -> PaintedLayer:
+    """Individual glyf entries, including composites, can be reused.
+
+    COLR lets us reuse the shape regardless of paint so paint is not part of key."""
+    return painted_layer._replace(paint=PaintSolid())
+
+
+def _ufo_colr_layers(colr_version, colors, color_glyph, glyph_cache):
+    # The value for a COLOR_LAYERS_KEY entry per
+    # https://github.com/googlefonts/ufo2ft/pull/359
+    colr_layers = []
+
+    # accumulate layers in z-order
+    layer_idx = 0
+    while layer_idx < len(color_glyph.painted_layers):
+        painted_layer = color_glyph.painted_layers[layer_idx]
+        layer_idx += 1
+
+        # if we've seen this shape before reuse it
+        glyph_cache_key = _inter_glyph_reuse_key(painted_layer)
+        if glyph_cache_key not in glyph_cache:
+            glyph = _create_glyph(color_glyph, painted_layer)
+            glyph_cache[glyph_cache_key] = glyph
+        else:
+            glyph = glyph_cache[glyph_cache_key]
+
+        layer = _colr_layer(colr_version, glyph.name, painted_layer.paint, colors)
+
+        colr_layers.append(layer)
+
+    if colr_version > 0 and len(colr_layers) > MAX_LAYER_V1_COUNT:
+        logging.info(
+            "%s contains > {MAX_LAYER_V1_COUNT} layers (%s); "
+            "merging last layers in PaintComposite tree",
+            color_glyph.glyph_name,
+            len(colr_layers),
+        )
+
+        i = MAX_LAYER_V1_COUNT - 1
+        colr_layers[i:] = [_build_composite_layer(colr_layers[i:])]
+
+    return colr_layers
+
+
+def _build_composite_layer(layers: Sequence[Mapping[str, Any]]) -> Mapping[str, Any]:
+    """Construct PaintComposite binary tree from list of UFO color layers.
```

Comment:
I guess now we could use a flatter tree of PaintColrLayers. Maybe in a separate PR if you prefer.

Reply:
+1 to flatter and separate PR